public/app/features/datasources/components/picker/DataSourceDropdown.test.tsx

```diff
@@ -0,0 +1,125 @@
+import { render, screen } from '@testing-library/react';
+import userEvent from '@testing-library/user-event';
+import React from 'react';
+
+import { DataSourceDropdown } from './DataSourceDropdown';
+
+const mockDS = {
+  name: 'mockDS',
+  uid: 'mockDSuid',
+  meta: { name: 'mockDS', info: { logos: { small: 'mockLogoPath' } } },
+};
+
+const xMockDS = {
+  name: 'xMockDS',
+  uid: 'xMockDSuid',
+  meta: { name: 'xMockDS', info: { logos: { small: 'xMockLogoPath' } } },
+};
+
+const mockDSList = [mockDS, xMockDS];
+
+const setup = (onChange = () => {}) => {
+  const props = { onChange, current: mockDS.name };
+  window.HTMLElement.prototype.scrollIntoView = function () {};
+  return render(<DataSourceDropdown {...props}></DataSourceDropdown>);
+};
+
+jest.mock('@grafana/runtime', () => {
+  const actual = jest.requireActual('@grafana/runtime');
+  return {
+    ...actual,
+    getTemplateSrv: () => {
+      return {
+        getVariables: () => [],
+      };
+    },
+  };
+});
+
+jest.mock('@grafana/runtime/src/services/dataSourceSrv', () => {
+  return {
+    getDataSourceSrv: () => ({
+      getList: () => mockDSList,
+      getInstanceSettings: () => mockDS,
+    }),
+  };
+});
+
+describe('DataSourceDropdown', () => {
```

Comment:
There are some important interactions we're not testing:
1. When the user clicks on a DS, the DS is selected, and the dropdown is closed
2. When the user clicks in open advanced, the modal is open
3. When the user clicks on Add CSV, the callback is called
4. When searching and not finding results, it shows the empty state with the CTA to add a new DS

Also, there are some configurations important for the components we're not testing either
1. The DS picker accepts `mixed`, `dashboard`, and many other booleans that are sent to the DS service to fetch the list. They are widely used for different products depending on the items we want to display in the dropdown.
2. When `current` is set, the value is shown at the top, and selected
3. When recently used + variables + builtIn + current are combined in the list, they should follow a specific order.

Reply:
These tests are all added in https://github.com/grafana/grafana/pull/68236/commits/193100a86fc956bbec04e168d0410bc12cbb3c6e

For the ordering I've opted to test that we use the correct parameters to call `getDataSourceCompareFn` since the sorting is tested in `utils.test.ts`

Testing that the filtering works as expected is done by testing that the correct parameters are sent to `dataSourceSrv.getList` testing that the actual filtering works is done in `datasource_srv.test.ts`

Reply:
Nice! 💯